RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,336 @@
+# Unit checking
+
+Unit checking is the process of inferring the units of Modelica variables, and ensuring the variables are used consistently in expressions and equations.
+
+It takes place after any evaluation of evaluable parameters and no evaluable parameters shall be evaluated during this process.
+(This must be true as long as tools allow the users to opt-out from unit-checking.)
+
+It is performed by using a type inference algorithm, albeit adapted to work with units.
+The algorithm proposed here is based on the Hindley-Milner algorithm.
+At a high level the algorithm collects a set of constraints on the units of expressions.
+It extracts a set of substitutions from these constraints which once applied determine whether a variable has a unit and what that unit would be.
+Any inconsistency discovered during this process is a unit error, and any unit expression that contains remaining unknowns at the end of the process is considered unknown.
+
+The document is organized as follows:
+* [Introduction of what is meant by the unit of a variable](#the-unit-of-a-modelica-variable)
+* [Description of the objects used in the inference algorithm](#unit-constraints-and-meta-expressions)
+* [Definition of unit equivalence and convertibility](#unit-equivalence-and-convertibility)
+* [Unit inference algorithm](#unit-inference)
+* [Application to Modelica](#unit-constraints-of-a-modelica-model)
+* [Modelica operators on units](#modelica-operators-on-units)
+* [Extensions and limitations](#possible-extensions-and-current-limits)
+
+## The unit of a Modelica variable
+
+The unit of a Modelica variable is the unit associated with the variable after unit-checking has been performed, possibly remaining unknown.
+
+Note that the unit of a variable is not the same concept as the `unit`-attribute of the variable, but in a unit-consistent model, the unit will be equal to the `unit`-attribute when present.
```

Comment:
Hans: Prefer something like.
If the variable has a unit-attribute that will be unit of the variable.
Otherwise it may be derived.